### DISCUSSION TITLE
Any way to condition the return type of a method, based on the return value of another method?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#8259] [Q&A] [answered]
Consider the following example:
```php
class ItemContainer
{
	public function hasNew() : bool
	{
		/* ... */
	}

	public function getNew() : ?SomeItemType
	{
		/* ... */
	}
}
```
What I would like to be able to do, is to tell Psalm that in the following code, `getNew()`, can only return `SomeItemType` (not nullable):
```php
if ($container->hasNew())
	$container->getNew()->useTheItem();	// Would like no error here, related to the nullable type.
```

Is this possible?

**Comment by AndrolGenhald** [comment, marked as answer]
Method calls can be asserted when [memoizeMethodCallResults](https://psalm.dev/docs/running_psalm/configuration/#memoizemethodcallresults) is `true` or when the methods are [`@psalm-mutation-free`](https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-mutation-free), but you can't call any possibly mutating methods between the `has()` call and the `get()` call: https://psalm.dev/r/e33bfb35a9

**Reply by rzvc** [reply]
Thanks, this is great.

**Reply by rzvc** [reply]
Looks like it doesn't work with aliases of imported types:

https://psalm.dev/r/410c4e737a
